tests/messengers/telegram/test_types.py

```diff
@@ -0,0 +1,231 @@
+@pytest.mark.skipif(not TG_API_ID or not TG_API_HASH, reason="TG credentials missing")
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    ["ui", "markup_type", "button_type"],
+    [
+        (
+            Keyboard(
+                buttons=[
+                    Button(text="button 1", payload=json.dumps({"text": "1", "other_prop": "4"})),
+                    Button(text="button 2", payload=json.dumps({"text": "2", "other_prop": "3"})),
+                    Button(text="button 3", payload=json.dumps({"text": "3", "other_prop": "2"})),
+                    Button(text="button 4", payload=json.dumps({"text": "4", "other_prop": "1"})),
+                ]
+            ),
+            types.InlineKeyboardMarkup,
+            types.InlineKeyboardButton,
+        ),
+        (
+            Keyboard(
+                is_inline=False,
+                buttons=[
+                    Button(text="button 1", payload=json.dumps({"text": "1", "other_prop": "4"})),
+                    Button(text="button 2", payload=json.dumps({"text": "2", "other_prop": "3"})),
+                    Button(text="button 3", payload=json.dumps({"text": "3", "other_prop": "2"})),
+                    Button(text="button 4", payload=json.dumps({"text": "4", "other_prop": "1"})),
+                ],
+            ),
+            types.ReplyKeyboardMarkup,
+            types.KeyboardButton,
+        ),
+    ],
+)
+async def test_buttons(ui, button_type, markup_type, tg_client, basic_bot, user_id, bot_id):
+    generic_response = Response(text="test", ui=ui)
+    telegram_response = TelegramResponse.parse_obj(generic_response)
+    assert telegram_response.text == generic_response.text
+    assert telegram_response.ui and isinstance(telegram_response.ui.keyboard, markup_type)
+    print(telegram_response.ui.keyboard.keyboard)
+    assert len(telegram_response.ui.keyboard.keyboard) == 2
+    reiterate_response(basic_bot, user_id, telegram_response)
+    minute_ago = datetime.datetime.now() - datetime.timedelta(minutes=1)
+    messages = await tg_client.get_messages(bot_id, None, offset_date=minute_ago)
+    assert messages
+
+
+@pytest.mark.skipif(not TG_BOT_TOKEN, reason="`TG_BOT_TOKEN` missing")
+@pytest.mark.skipif(not TG_API_ID or not TG_API_HASH, reason="TG credentials missing")
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    ["ui"],
+    [
+        (TelegramUI(keyboard=types.ReplyKeyboardRemove()),),
+    ],
+)
+async def test_keyboard_remove(ui, basic_bot, user_id, tg_client, bot_id):
+    generic_response = Response(text="test", ui=ui)
+    telegram_response = TelegramResponse.parse_obj(generic_response)
+    assert telegram_response.text == generic_response.text
+    assert telegram_response.ui
+    reiterate_response(basic_bot, user_id, telegram_response)
+    minute_ago = datetime.datetime.now() - datetime.timedelta(minutes=1)
+    messages = await tg_client.get_messages(bot_id, None, offset_date=minute_ago)
+    assert messages
```

Comment:
Maybe assert that they are the same as the ones the bot sent?

Reply:
Done